Tutorials/13_Geometry-Optimization/opt_helper/v3d.py

```diff
@@ -0,0 +1,171 @@
+# Functions to compute properties of 3d vectors, including angles,
+# torsions, out-of-plane angles.  Several return False if the operation
+# cannot be completed numerically, as for example a torsion in which 3
+# points are collinear.
+import numpy as np
+from math import sqrt, fabs, sin, acos, asin, fsum
+
+TORS_ANGLE_LIM = 0.017
+TORS_COS_TOL   = 1e-10
+DOT_PARALLEL_LIMIT = 1.e-10
+
+def norm(v):
+    return sqrt(v[0] * v[0] + v[1] * v[1] + v[2] * v[2])
+
+
+def dot(v1, v2, length=None):
+    if length is None:
+        return v1[0] * v2[0] + v1[1] * v2[1] + v1[2] * v2[2]
+    else:
+        return fsum([v1[i] * v2[i] for i in range(length)])
+
+
+def dist(v1, v2):
+    return sqrt((v2[0] - v1[0])**2 + (v2[1] - v1[1])**2 + (v2[2] - v1[2])**2)
+
+
+# Normalize vector in place.  If norm exceeds thresholds, don't normalize and return False..
+def normalize(v1, Rmin=1.0e-8, Rmax=1.0e15):
+    n = norm(v1)
+    if n < Rmin or n > Rmax:
+        return False
+    else:
+        v1 /= n
+        return True
+
+
+def axpy(a, X, Y):
+    Z = np.zeros(Y.shape, float)
+    Z = a * X + Y
+    return Z
```

Comment:
If `a` is scalar (usual daxpy), this is just `a * X + Y` in calling code, no need for fn.